### PR TITLE
SQL Server native delayed delivery

### PR DESCRIPTION
**Pull request by weralabaj**
relates to https://github.com/Particular/PlatformDevelopment/issues/1306

### CODE REVIEW

**Comment by weralabaj** [comment]
@SzymonPobiega I think with snippets we need to release to myget?

**Comment by SzymonPobiega** [comment]
I removed the upgrade bits so that it can be merged and then the upgrade PR can be merged after this one (#2937)

**Comment by SzymonPobiega** [comment]
@Particular/docs-maintainers @MarcinHoppe ready for review.

**Comment by SzymonPobiega** [comment]
ping.

File: transports/sql/native-delayed-delivery.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+versions: '[3.1,)'
+---
+
+In Versions 3.1 and above, the SQL Server transport can take advantage of native [delayed delivery](/nservicebus/messaging/delayed-delivery.md) without then need to run the [timeout manager](/nservicebus/messaging/timeout-manager.md). Instead, the transport creates infrastructure which can delay messages using native SQL Server transport features. To enable the native delayed delivery use the following API:
+
+snippet: EnableNativeDelayedDelivery
+
+NOTE: In this mode the timeout manager will still be running in order to process all outstanding delayed messages. Refer to the [Disabling the timeout manager](/transports/sql/native-delayed-delivery.md#disabling-the-timeout-manager) section for details on how to disable the timeout manager entirely.
+
+## How it works
```
Comment (weralabaj) [inline transports/sql/native-delayed-delivery.md]: Probably better as "configuration" header? I'd also link from here to the upgrade guide, but that can be done in the other PR (let's just make a note to not forget).

**Comment by weralabaj** [comment]
@SzymonPobiega One comment ;)

File: transports/sql/native-delayed-delivery.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+In Versions 3.1 and above, the SQL Server transport can take advantage of native [delayed delivery](/nservicebus/messaging/delayed-delivery.md) without then need to run the [timeout manager](/nservicebus/messaging/timeout-manager.md). Instead, the transport creates infrastructure which can delay messages using native SQL Server transport features. To enable the native delayed delivery use the following API:
+
+snippet: EnableNativeDelayedDelivery
+
+NOTE: In this mode the timeout manager will still be running in order to process all outstanding delayed messages. Refer to the [Disabling the timeout manager](/transports/sql/native-delayed-delivery.md#backwards-compatibility-disabling-the-timeout-manager) section for details on how to disable the timeout manager entirely.
+
+## How it works
+
+The transport creates an additional queue that stores delayed messages. The table name has the format `endpoint-name.suffix`, using the suffix specified in the configuration:
```
Comment (MarcinHoppe) [inline transports/sql/native-delayed-delivery.md]: When reading this, I am not sure if these are mandatory settings. I think we should be explicit about what is required and which settings have a default value.

File: transports/sql/native-delayed-delivery.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+snippet: DelayedDeliveryTableSuffix
+
+The timeouts poller checks for expired messages at specified interval:
```
Comment (MarcinHoppe) [inline transports/sql/native-delayed-delivery.md]: > The timeouts poller checks for expired messages

would be better as

> The transport polls for expired messages...

File: transports/sql/native-delayed-delivery.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+snippet: DelayedDeliveryProcessingInterval
+
+The poller picks and dispatches messages in batches of specified size:
```
Comment (MarcinHoppe) [inline transports/sql/native-delayed-delivery.md]: Ditto. The poller is an implementation detail and it does not have to exist in our user's minds.

File: transports/sql/native-delayed-delivery.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+snippet: DelayedDeliveryBatchSize
+
+When the delay time lapses, SQL Server transport moves a batch of messages to the input queue. Note that this means the exact time of delivering delayed message is always approximate.
```
Comment (MarcinHoppe) [inline transports/sql/native-delayed-delivery.md]: input queue -> destination queue?

File: transports/sql/native-delayed-delivery.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+
+## Backwards compatibility
+
+When upgrading to a version of the transport that supports native delayed delivery, it is safe to operate a combination of endpoints using native delayed delivery and endpoints using persistence-based delayed delivery at the same time:
```
Comment (MarcinHoppe) [inline transports/sql/native-delayed-delivery.md]: persistence-based delayed delivery -> timeout manager?

File: transports/sql/native-delayed-delivery.md (hunk 4 above)
Comment (weralabaj) [inline transports/sql/native-delayed-delivery.md]: Given that they have to configure it, I wouldn't say it's just an implementation detail. Also, I'm not sure how we could explain how it works without mentioning the poller.

File: transports/sql/native-delayed-delivery.md (hunk 2 above)
Comment (weralabaj) [inline transports/sql/native-delayed-delivery.md]: Seems all are now required, correct @SzymonPobiega ?

File: transports/sql/native-delayed-delivery.md (hunk 4 above)
Comment (MarcinHoppe) [inline transports/sql/native-delayed-delivery.md]: Then the poller should be introduced earlier on and an explanation should be given about what it is and how it works. Also, for non-optional params we should provide guidance on how to set them.

How do I know what the right interval and batch size is?

Comment (weralabaj) [inline transports/sql/native-delayed-delivery.md]: > Then the poller should be introduced earlier on and an explanation should be given about what it is and how it works.

We don't do that in any other place where we do polling, AFAIK. If we say instead "SQL Transport polls messages at intervals specified by..." then would that be acceptable?

> for non-optional params we should provide guidance on how to set them.

OK. Also came to my mind but all I could think about is giving an example, as we don't have any specific "perf tuning" techniques here I think.

Comment (MarcinHoppe) [inline transports/sql/native-delayed-delivery.md]: I [think that would be better](https://github.com/Particular/docs.particular.net/pull/2815#discussion_r126355399).

Comment (weralabaj) [inline transports/sql/native-delayed-delivery.md]: OK, misunderstood your comment ;)

Comment (MarcinHoppe) [inline transports/sql/native-delayed-delivery.md]: If we give an example, then I think we should explain the thinking that went into choosing these values, preferably connecting to some business-level requirements around how accurate we need to be with the timeouts and the throughput maybe.

**Comment by weralabaj** [comment]
@MarcinHoppe please, see now :)

File: transports/sql/native-delayed-delivery.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+In Versions 3.1 and above, the SQL Server transport can take advantage of native [delayed delivery](/nservicebus/messaging/delayed-delivery.md) without then need to run the [timeout manager](/nservicebus/messaging/timeout-manager.md). Instead, the transport creates infrastructure which can delay messages using native SQL Server transport features. To enable the native delayed delivery use the following API:
+
+snippet: EnableNativeDelayedDelivery
+
+NOTE: In this mode the timeout manager will still be running in order to process all outstanding delayed messages. Refer to the [Disabling the timeout manager](/transports/sql/native-delayed-delivery.md#backwards-compatibility-disabling-the-timeout-manager) section for details on how to disable the timeout manager entirely.
+
+
+## Configuration
+
+The settings described in this section are obligatory. By default use the values presented in snippets. The settings can be fine-tuned to fit a particular system's characteristic, e.g. in case the expired timeout messages need to be picked up more frequently to result in a more precise approximation of expiry time.
```
Comment (MarcinHoppe) [inline transports/sql/native-delayed-delivery.md]: If we recommend:

> By default use the values presented in snippets.

Then why they are not default in the code?

File: transports/sql/native-delayed-delivery.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+The transport creates an additional queue that stores delayed messages. The table name has the format _`endpoint-name.suffix`_, using the suffix specified in the configuration:
```
Comment (MarcinHoppe) [inline transports/sql/native-delayed-delivery.md]: Technically speaking it is not a queue table. Maybe let's just state that it creates an additional table?

File: transports/sql/native-delayed-delivery.md (hunk 7 above)
Comment (SzymonPobiega) [inline transports/sql/native-delayed-delivery.md]: I think they are code-defaults, @weralabaj ? I don't think we require any of these.

File: transports/sql/native-delayed-delivery.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+snippet: DelayedDeliveryBatchSize
+
+When the delay time lapses, SQL Server transport moves a batch of messages to the destination queue, i.e. the endpoint's input queue. Note that this means the exact time of delivering delayed message is always approximate.
+
+
+## Backwards compatibility
+
+When upgrading to a version of the transport that supports native delayed delivery, it is safe to operate a combination of endpoints using native delayed delivery and endpoints using timeout manager at the same time:
+- Endpoints with native delayed delivery can send delayed messages to endpoints using persistence-based delayed delivery.
```
Comment (MarcinHoppe) [inline transports/sql/native-delayed-delivery.md]: persistence-based delayed delivery -> timeout manager.

File: transports/sql/native-delayed-delivery.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+snippet: DelayedDeliveryBatchSize
+
+When the delay time lapses, SQL Server transport moves a batch of messages to the destination queue, i.e. the endpoint's input queue. Note that this means the exact time of delivering delayed message is always approximate.
+
+
+## Backwards compatibility
+
+When upgrading to a version of the transport that supports native delayed delivery, it is safe to operate a combination of endpoints using native delayed delivery and endpoints using timeout manager at the same time:
+- Endpoints with native delayed delivery can send delayed messages to endpoints using persistence-based delayed delivery. 
+- Endpoints with native delayed delivery can continue to receive delayed messages from endpoints using persistence-based delayed delivery.
```
Comment (MarcinHoppe) [inline transports/sql/native-delayed-delivery.md]: persistence-based delayed delivery -> timeout manager.

**Comment by MarcinHoppe** [comment]
@weralabaj Thanks, looks much better now. Last batch of (minor) comments.

File: transports/sql/native-delayed-delivery.md (hunk 7 above)
Comment (weralabaj) [inline transports/sql/native-delayed-delivery.md]: My mistake, MatureBatchSize and Suffix have defaults, I'll update.
@SzymonPobiega I may be wrong but I don't see where the default for interval is specified, although in the comment we say "Configures how often delayed messages are processed (every 5 seconds by default)."

**Comment by weralabaj** [comment]
The only outstanding issue is to clarify if interval setting should have a default value (and does have it). Comment says one thing, but I can't find the relevant code ;)

**Comment by SzymonPobiega** [comment]
Well, the default interval was `default(TimeSpan)` :joy:. That actually explains why my machine was hanging where there was a problem with DB. It kept running the query in a tight loop.

I've changed it [here](https://github.com/Particular/NServiceBus.SqlServer/commit/be9f5fc7eb3fbef4fb126b8a0365c4803f155c2c) to 1 second to be aligned with Timeout Manager. I've also updated the snippets.

**Comment by weralabaj** [comment]
@MarcinHoppe should be ok now I think :)

**Pull request merged**